### ISSUE TITLE
[JSX] change default set of JSX equivalences.

### ISSUE THREAD

**Issue by aryx** [semgrep/semgrep#2812]
Right now in semgrep, the pattern <$EL /> will match lots of JSX elements: <foo />, but also <foo ></foo>, also <foo>xxx</foo>, etc.
and internally a pattern like <$EL></$EL> is actually transformed in <$EL />. This is maybe too much because it prevents users like @yoav-lavi to explicitely match code like <div></div> but not <div />.

@minusworld @mschwager @ievans maybe we should disable most of those equivalences. That means we will have to rewrite some of our semgrep rules like https://github.com/returntocorp/semgrep-rules/blob/develop/typescript/react/security/audit/react-css-injection.yaml to use instead of patterns like <$EL style={$STYLE} />
the more verbose <$EL style={$STYLE} >...</$EL>.
What do you think? We should even maybe require to use ... for the attributes to says more attributes are ok (to be consistent with what we do elsewhere @mschwager ?)
Then the only builtin equivalence we can still do is that <$EL ...>...<$EL> would still match code like <foo></foo> or <foo />.
Feedback?

**Comment by aryx** [comment]
origin of the conversation here: https://github.com/returntocorp/semgrep/issues/2685

**Comment by aryx** [comment]
another related pb mentioned by gitlab: https://semgrep.dev/s/squaresurf:react-dangerouslysetinnerhtml
link to slack: https://r2c-community.slack.com/archives/C01KY32FDAT/p1615507626047700

**Comment by mschwager** [comment]
> use instead of patterns like <$EL style={$STYLE} />
the more verbose <$EL style={$STYLE} >...</$EL>.
What do you think? We should even maybe require to use ... for the attributes to says more attributes are ok (to be consistent with what we do elsewhere @mschwager ?)

That sounds good to me. We may even want to consider the opposite equivalence: `<$EL style={$STYLE}>...</$EL>` matches `<$EL style={$STYLE} />`. In this case we can always remove special cases with a `pattern-not`. Generally I find that more verbose patterns catching more things is useful because we can remove special cases if necessary.

**Comment by aryx** [comment]
You're right! Ok, I'll modify the engine but we will have to modify a few semgrep-rules then.

**Comment by aryx** [comment]
What about the attributes? should we require ... in the list of attributes to say more are ok? Or have it implicit like we do inside objects or classes?

**Comment by mschwager** [comment]
> What about the attributes? should we require ... in the list of attributes to say more are ok? Or have it implicit like we do inside objects or classes?

How does it work inside objects or classes?

My first thought is we could consider attributes like function parameters. That is, we don't include an implicit `...` - the following all have different meanings: `func()`, `func(...)`, `func(..., foo="bar", ...)`, etc. What do you think?

**Comment by aryx** [comment]
Inside object or classes right now you don't have to put '...', which may not be right.

I think it's better to require '...' for attributes (and probably also for object/classes) to be more consistent.

**Comment by aryx** [comment]
:wave:

**Comment by brandonspark** [comment]
Is this fixed now since the addition of `xml_singleton_loose_matchin`?
